### ISSUE TITLE
Playlist import / export

### ISSUE THREAD

**Issue by antipiot** [deluan/navidrome#105]
Hello! First of all, thanks for your work! Looks very promising.

Actually using Airsonic.

I wanted to ask if it will be possible to have a playlist management (i saw smart / dynamic playlists on the road map) but i'd like to be able to manage standarf m3u / m3u8 playlists with the ability to import - export them as mine are very precious to me :-)

**Comment by deluan** [comment]
Thanks for trying Navidrome out. 

Playlist management will definitely comes before smart playlists. Although I have an issue open (#10 ) to track all initial work on the WebUI, I'll leave this open for eventual discussions specific about playlists.

**Comment by antipiot** [comment]
Thanks to you!
Lookings forward for the playlists / smart playlists and vignette enhancement.
Will probably switch once this is done. 
really appreciate the docker container aswell.

Anyway, many thanks again!

**Comment by legobuild** [comment]
Just found Navidrome and it looks like a great lightweight replacement for Airsonic. Looks like I can create playlists in play:sub, just need the management capabilities in the UI. Keep up the great work!

**Comment by arg274** [comment]
Owing to the discussion on Discord, I also think that playlists should be kept as-is even if the M3U it was originally imported from is not found in its location. Deletion should only be handled by the user explicitly.

**Comment by git1873** [comment]
Adding to the idea of @arg274, it would be an easy way to make the playlist feature open to other apps if Navidrome synced the changes made back to the playlist file (M3U or else). That way one could add a track to one playlist in Navidrome and have the changes automatically visible in a standalone software like MusicBee, iTunes or DJ Software.

**Comment by deluan** [comment]
@git1873 two-way sync is really hard to get right. What to do in case of conflicts (changes on both playlists being synched)? 

I think the best approach is one way sync: Navidrome will import m3u/pls files, and will flag it as "synchronized". The imported playlist will be read-only in Navidrome, but if the file it was imported from changes, it will update. If you want to disable the sync, you'd disable the playlist's "synchronized" flag (in the UI)

**Comment by Tontonjo** [comment]
On my side, i would never use the folder to import playlists into Navidrome as it is my main player but will defininitly use it to export the playlists i've made using GUI / Dsub to another players, like Moode-audio.

**Comment by antipiot** [comment]
On my side, i would never use the folder to import playlists into Navidrome as it is my main player but will defininitly use it to export the playlists i've made using GUI / Dsub to another players, like Moode-audio.

**Comment by git1873** [comment]
@deluan You are right, you can't really build a rock-solid 2-way sync solution on a file based system without locking/transactions and conflict management in clients. I am not even sure how many users would use the feature, but i certainly would. 
If only one user is accessing the same set of playlist files from different devices, getting it to sync is not really hard. Navidrome could fetch the current state of the playlist file before displaying it to the user (and before adding something to it). This way, no conflicts would occur, unless two clients write at (or depending on the other clients behaviour load, resuting in later conflicts because client cached an older version of the file) the same time. Not a "set up and forget" solution for every user, it would just make some workflows possible.

Aside from that; the one-way sync option you proposed is still a good idea and a good thing to start with. Playlists could be configured as Navidrome either being Master oder Slave. If Slave, Navidrome can't write to a playlist, only sync its contents from a file. If Master, Navidrome can edit the Playlist and write it to a file (but not read it back).

**Comment by certuna** [comment]
Testing out the playlists - am I correct that with playlists so far in v0.19 you can:
- add individual tracks from the Tracks view
and you can not yet:
- add individual tracks from the Albums view
- add whole albums from the Albums view
- change the playlist order
- sort the playlist order
- see other fields than title/artist/time

**Comment by deluan** [comment]
Hey @certuna, from the comment you added in the PR #265, the only items not implemented yet are:
> - change the playlist order
> - see other fields than title/artist/time

All other items are already in 0.19.0 and/or in the develop build right now. Let me know if you are having trouble accessing them.

Question: What other fields do you want to see in the playlist tracks grid? Keep in mind that you can expand each row, which gives you more info about each song

**Comment by certuna** [comment]
Yes this all works well, now testing 0.21, I love how fast all this development is going.

I would imagine you'd want the same columns in Playlists as you have in the rest of the library, ie that you can choose to view a Playlist in Album view, Song view, Artist view.

**Comment by deluan** [comment]
Ah, I got it. Yes I want to have different visualization modes for playlists, but right now there are a lot of other things that would come first, ex: smart playlists. When implementing the artist view, I may refactor all views to allow them to be reused with playlists. I'll add this to the issue's checklist, even though it will probably move to its own issue once the other items in the list are done

**Comment by Extarys** [comment]
Eventually: Share playlist with a user, or allow a user on the server to check other's playlists and "follow" them

**Comment by deluan** [comment]
@Extarys Playlists can already be set as public, then all users in a server have access to it. Is it enough?

**Comment by deluan** [comment]
Update: Latest from master implements the playlist import. Just drop your `.m3u` files in any directory under your music folder, and Navidrome will parse and import it automatically on the next scan. Playlists that were imported this way will be kept in sync: Any changes to the `.m3u` file will update the Playlist. If you want to disable this sync after the playlist has been imported, you can edit it in the UI and disable it ("Auto-Import" switch).

To be able to use it, you'll need to get the latest `develop` build, and set the `DevNewScanner=true` config option.

**Comment by antipiot** [comment]
Hi Deluan!
This looks perfect.

For people who want to try in docker pull navidrome:develop and add ND_DEVNEWSCANNER=true environement variable.

Tried it myself and my playlist are not imported ATM because of an error:
`
2020/07/18 12:11:42 OK    20200411164603_add_created_and_updated_fields_to_playlists.go
2020/07/18 12:11:42 OK    20200516140647_add_playlist_tracks_table.go
2020/07/18 12:11:42 OK    20200710211442_add_playlist_path.go
time="2020-07-18T12:11:44+02:00" level=error msg="Error retrieving playlist owner" error="data not found"`
Do i need to edit my playlists in any way?

FYI:
My playist is named:
"instrumental Metal"

and entries looks like:
`/music/Andy James/Exodus/Andy James - Never Back Down.mp3
/music/Wayne Calford/The Guitar Made Me Do It/Wayne Calford - Devil's Groove.mp3
/music/Marty Friedman/Dragon's Kiss/Marty Friedman - Dragon Mistress.mp3
/music/Steve Vai/The Story of Light/Steve Vai - Velorum.mp3
/music/Angel Vivaldi/Universal Language/Angel Vivaldi - A Mercurian Summer.mp3
/music/Wayne Calford/Remembering Never/Wayne Calford - The End of December.mp3`

Any hint?

**Comment by deluan** [comment]
@antipiot Thanks for helping testing this. Seems the issue is related to your users. You have at least one admin user, right?

Can you change the loglevel to `trace` and try again?

**Comment by antipiot** [comment]
@deluan:
This happened when i started a brand new container  - thus the error reporting no admin may have been right.
I've tested again with a test container, created an admin - stopped the container then setted the new newscanner env variable to true but playlist sill aren't imported.

Do i miss something?

![image](https://user-images.githubusercontent.com/26579988/87860264-56d14f00-c93c-11ea-893a-9a66f2f584b3.png)

**Comment by antipiot** [comment]
Any chance to have the changes wrote in the m3u file?

Let me clarify my needs:
I'm using a script to automatically grab my playlists in a folder and "convert" them to fit in my Moode-Audio player.

I would like to generate and edit my playlists in Navidrome and be able to grab them with ease.

**Comment by deluan** [comment]
Export playlists come next, I've been thinking of implementing something similar to what @git1873 described in his comment:
> If Master, Navidrome can edit the Playlist and write it to a file (but not read it back).

You could set a path in the Playlist Edit UI, so everytime it was changed, it auto-exports to that path... 

If you don't want to wait, you could use Subsonic API with `curl` and parse the results with `jq` to grab the playlists directly from Navidrome.

**Comment by antipiot** [comment]
@deluan i did tried something to download playlist with m3u format on my Airsonic but couldn't get it working efficiently due to some mechanisms implied with auth token or something. Read somewhere i wasn't alone trying and facing a wall on the case.

"You could set a path in the Playlist Edit UI, so everytime it was changed, it auto-exports to that path..."
How may i set a custom path? 
![image](https://user-images.githubusercontent.com/26579988/87862023-f813d180-c94b-11ea-93ae-35c506e73c65.png)

Thanks for you great work on Navidrome!

**Comment by deluan** [comment]
> @deluan i did tried something to download playlist with m3u format on my Airsonic but couldn't get it working efficiently due to some mechanisms implied with auth token or something. Read somewhere i wasn't alone trying and facing a wall on the case.

Yes, that is hard. What I'm proposing is something different: to get the playlist in JSON, using the API and converting it to m3u in your script. If you want to try and need help, join our Discord server and I can help you with that (it is easier to chat than exchange replies here... :) )

Re: Path, in the currently implementation, you can't change the path. It is there just to let you know the origin of the imported playlist. But when implementing auto-export, this path will be configurable.

**Comment by antipiot** [comment]
Got it!
Thanks for the clear explanation - will wait for the implementation patiently. 
no need to script further if something covering the need will be available.

**Comment by antipiot** [comment]
@deluan ,sorry to bother you:
If i'm not mistaken your the new scanner has been merged.
This scan - import should appear in logs? asking because my playlists aren't imported.
did i miss something?

**Comment by deluan** [comment]
Yes, from version 0.26.1, it is the default scanner. 

In default `info` loglevel, it should show a message like this when it finds a new or changed `.m3u` file:
```
Adding synced playlist      owner=admin path="/Users/deluan/Music/Music/Media/Music/Best Of Saara Saara.m3u" playlist="Best Of Saara Saara"
```

You can try making a simple change to the playlist (adding a comment or something like this), and once you save it, it should picked-up in the next scan cycle. If not, repeat the process with loglevel = `trace` and send me the logs to try to figure out why it is not working.

**Comment by antipiot** [comment]
@deluan , Thanks for your answer:
I do not know why but i had to remove the playlists from media folder, reboot the server, put the playlists back in and they where imported correctly.

Thanks for your great work and help!

**Comment by deluan** [comment]
@antipiot That's very weird. Just updating the playlists without any reboot should be enough. Well, if you can reproduce this behavior consistently, please open a new GH issue with steps on how to reproduce it, so we can discuss try to figure this out

**Comment by andrewzah** [comment]
@deluan can the importing of m3u's be made optional? For cd ripping with EAC, it's standard to include a .m3u of the album.

**Comment by deluan** [comment]
Hummm.... I'm not actually a fan of (and don't use any) `.m3u` playlists, but I can see disabling the auto-import being desired. Sure, I'll add an option in the next release

**Comment by deluan** [comment]
@andrewzah Done in 28d1428

**Comment by Merkwurdichliebe** [comment]
Until such a command is available in the UI, Is there any SQL command I can use to mass export all playlists from the navidrome database as `m3u8` files, as a backup in case something goes wrong?

**Comment by deluan** [comment]
@Merkwurdichliebe, please open a new issue with your request, or if it is just a question feel free to ask in Discord. This issue one was completed 2 years ago.
